File: Baekjoon/Baekjoon02156.cpp
// Baekjoon02156.cpp
#include <iostream>

using namespace std;

int main(void) {
    ios_base::sync_with_stdio(false);
    cin.tie(0);

    int n;
    cin >> n;

    int wine[10'000];
    for (int i = 0; i < n; i++)
        cin >> wine[i];

    int ans = wine[0];
    int dp[3][10'000] = { { 0, }, };
    dp[0][0] = 0, dp[1][0] = wine[0], dp[2][0] = wine[0];
    for (int i = 1; i < n; i++) {
        dp[0][i] = max(dp[0][i - 1], max(dp[1][i - 1], dp[2][i - 1]));
        dp[1][i] = dp[0][i - 1] + wine[i];
        dp[2][i] = dp[1][i - 1] + wine[i];

        ans = max(max(ans, dp[0][i]), max(dp[1][i], dp[2][i]));
    }

    cout << ans << '\n';

    return 0;
}

File: Baekjoon/Baekjoon02156.py
# Baekjoon02156.py
n = int(input())
wine = [int(input()) for _ in range(n)]

dp = [
    [0 for _ in range(n)],
    [0 for _ in range(n)]
]
dp[0][0] = dp[1][0] = wine[0]

ans = wine[0]
if n > 1:
    dp[0][1] = wine[1]
    dp[1][1] = wine[0] + wine[1]
    ans = max(ans, dp[0][1], dp[1][1])

    for i in range(2, n):
        if i > 2:
            dp[0][i] = max(dp[0][i - 2], dp[1][i - 2], dp[0][i - 3], dp[1][i - 3]) + wine[i]
        else:
            dp[0][i] = max(dp[0][i - 2], dp[1][i - 2]) + wine[i]
        dp[1][i] = dp[0][i - 1] + wine[i]

        ans = max(ans, dp[0][i], dp[1][i])

print(ans)
